src/ristretto/ristretto_keys.rs

```diff
@@ -74,29 +74,49 @@ impl borsh::BorshDeserialize for RistrettoSecretKey {
 //-----------------------------------------   Ristretto Secret Key    ------------------------------------------------//
 impl SecretKey for RistrettoSecretKey {
     const KEY_LEN: usize = 32;
+    const WIDE_REDUCTION_LEN: usize = 64;
 
     /// Return a random secret key on the `ristretto255` curve using the supplied CSPRNG.
     fn random<R: RngCore + CryptoRng>(rng: &mut R) -> Self {
         RistrettoSecretKey(Scalar::random(rng))
     }
+
+    /// Return a secret key computed by wide reduction on a byte array
+    /// If the byte array is not exactly 64 bytes, returns an error
+    fn from_bytes_wide(bytes: &[u8]) -> Result<Self, ByteArrayError> {
+        if bytes.len() != Self::WIDE_REDUCTION_LEN {
+            return Err(ByteArrayError::IncorrectLength {});
+        }
+
+        let mut bytes_copied = Zeroizing::new([0u8; Self::WIDE_REDUCTION_LEN]);
+        bytes_copied.copy_from_slice(bytes);
+
+        Ok(RistrettoSecretKey(Scalar::from_bytes_mod_order_wide(&bytes_copied)))
+    }
 }
 
 //-------------------------------------  Ristretto Secret Key ByteArray  ---------------------------------------------//
 
 impl ByteArray for RistrettoSecretKey {
-    /// Create a secret key on the Ristretto255 curve using the given little-endian byte array. If the byte array is
-    /// not exactly 32 bytes long, `from_bytes` returns an error. This function is guaranteed to return a valid key
-    /// in the group since it performs a mod _l_ on the input.
+    /// Return a secret key computed from a canonical byte array
+    /// If the byte array is not exactly 32 bytes, returns an error
+    /// If the byte array does not represent a canonical encoding, returns an error
     fn from_bytes(bytes: &[u8]) -> Result<RistrettoSecretKey, ByteArrayError>
```

Comment:
I would prefer to not implement ByteArray and instead have a specific `from_bytes_cannonical` or `from_bytes_exact` trait

Reply:
Should other structs that have canonical representations (like Ristretto group elements) move to that trait as well?

Reply:
AFAIK the ByteArray trait makes it simple to use `to_hex()` but doesn't really have any other significant addition, apart from a marker. If I were doing the change I would probably just add the methods directly onto the structs without a trait. Beware that there are also Serialize and BorshSerialize methods which should all use the canonical versions and error if not canonical